### PR TITLE
Return the TLS write buffer when TCP client connection socket setup fails

### LINKED ISSUE #4828: [Audit-Medium] TCP exceptional construction paths do not release all acquired resources
> **AI-generated audit finding** — from a repository-wide audit performed by OpenAI Codex (`codex-cli`), independently verified against source by Claude (Anthropic), model **Opus 5** (`claude-opus-5`). It **has not been triaged by a human yet**; verify the reasoning, reproducibility, and severity before acting on it.

## Summary

`TcpClientConnection` rents the write-buffer owner in its **base-constructor initializer**, before the body runs:

```csharp
internal TcpClientConnection(...)
    : base(authenticationOptions is not null ? pool.Rent(minimumSegmentSize) : null)
{
    ...
    catch (SocketException exception) { Socket.Dispose(); throw exception.ToIceRpcException(); }
    catch { Socket.Dispose(); throw; }
}
```

Both `catch` blocks dispose the socket but not the rented owner. Because the constructor throws, no instance escapes, so `Dispose()` — which *would* return the buffer — is never called. The rented memory is never returned to the pool.

`TcpServerConnection` has the same base-initializer shape, and the listener path can similarly lose an accepted socket when server-connection construction or pool rent fails.

## Note

This is **not** the same as the "`Dispose()` returns its rented SSL write buffer on every call" claim that was refuted in the 2026-07-27 general-correctness audit. `Dispose()` is correct; the gap is the constructor failure path, where `Dispose()` never runs.

## Source references

- [src/IceRpc/Transports/Tcp/Internal/TcpConnection.cs#L309](https://github.com/icerpc/icerpc-csharp/blob/2de29310a82428560125c5f804127ea054f46769/src/IceRpc/Transports/Tcp/Internal/TcpConnection.cs#L309) — client rent in the base initializer
- [src/IceRpc/Transports/Tcp/Internal/TcpConnection.cs#L402](https://github.com/icerpc/icerpc-csharp/blob/2de29310a82428560125c5f804127ea054f46769/src/IceRpc/Transports/Tcp/Internal/TcpConnection.cs#L402) — server twin
- [src/IceRpc/Transports/Tcp/Internal/TcpConnection.cs#L323-L341](https://github.com/icerpc/icerpc-csharp/blob/2de29310a82428560125c5f804127ea054f46769/src/IceRpc/Transports/Tcp/Internal/TcpConnection.cs#L323) — catch blocks disposing only the socket

## Recommendation

Use explicit staged ownership with `try`/`catch` cleanup until construction succeeds and ownership transfers to the connection instance. Add fault-injection tests at each acquisition boundary.

## Audit context

M-31 of [repository-wide-audit-2026-07-27.md](https://github.com/icerpc/icerpc-csharp-audit/blob/main/repository-wide-audit-2026-07-27.md); confirmed by the [Claude verification pass](https://github.com/icerpc/icerpc-csharp-audit/blob/main/repository-wide-audit-claude-review-2026-07-28.md).

Severity: **Medium**.

### PR DESCRIPTION
**Pull request by pepone**
When the `TcpClientConnection` constructor failed to bind or configure its socket (for example a `LocalNetworkAddress`
the socket cannot bind), its two catch blocks disposed the socket but never returned the TLS write buffer rented in the
base-constructor initializer. Since the constructor throws, no instance escapes and `Dispose()`, which returns the
buffer, was never called.

The catch blocks now call `Dispose()` instead of `Socket.Dispose()`. At that point the SSL stream is still null and the
socket is not connected, so `Dispose()` closes the socket and returns the write buffer to the pool.

Fixes #4828

## What's Changed entry

None — the default pool doesn't leak, and a client socket setup failure means every connection attempt fails.

### CODE REVIEW

**Review by bernardnormier — COMMENTED** [review]
Let's keep the constructors as they are. For the client leak, calling `Dispose()` instead of `Socket.Dispose()` in
the two catch blocks of the `TcpClientConnection` constructor returns the buffer as well: the SSL stream is still
null and closing an unconnected socket is fine. Two lines, no `CreateSocket`, no base constructor change.

The server path needs a pool whose `Rent` throws, which the default pool never does for our segment sizes. Drop that
part.

No test for this one: the leak is only observable through a pool built to count rentals, and the fix is two lines.

No What's Changed entry either: `None — the default pool doesn't leak, and a client socket setup failure means every
connection attempt fails.`

**Comment by pepone** [comment]
🤖 **Claude:** Reworked as suggested in e391a32a67e84246b115f1931ab5da5cd6c821f6: the constructors are back as they were, the two catch blocks in the `TcpClientConnection` constructor now call `Dispose()` instead of `Socket.Dispose()`, and the server change, the tests, and the What's Changed entry are dropped.

File: src/IceRpc/Transports/Tcp/Internal/TcpConnection.cs [hunk 1]
```diff
@@ -337,12 +337,12 @@ internal TcpClientConnection(
         }
         catch (SocketException exception)
         {
-            Socket.Dispose();
+            Dispose();
```
[the remark this answers is not in the document]
Comment (bernardnormier) [reply src/IceRpc/Transports/Tcp/Internal/TcpConnection.cs:340]: Not covering those on purpose. `new Socket` fails only when the process can't create a socket at all, and with the default pool a dropped rental is garbage collected. The server path needs a pool whose `Rent` throws, which the default pool never does for our segment sizes. This is the fix we want for #4828.

**Review by bernardnormier — APPROVED** [review]

File: src/IceRpc/Transports/Tcp/Internal/TcpConnection.cs [hunk 2]
```diff
@@ -337,12 +337,12 @@ internal TcpClientConnection(
...
             throw exception.ToIceRpcException();
         }
         catch
         {
-            Socket.Dispose();
+            Dispose();
```
[the remark this answers is not in the document]
Comment (bernardnormier) [reply src/IceRpc/Transports/Tcp/Internal/TcpConnection.cs:345]: No test for this one: the leak is observable only through a pool built to count rentals, and the fix is two lines.

**Review by externl — APPROVED** [review]
